File: src/main/java/uk/co/rafearnold/randomworkoutroutine/util/RandomUtils.java
package uk.co.rafearnold.randomworkoutroutine.util;

import lombok.experimental.UtilityClass;

import java.util.Random;

@UtilityClass
public class RandomUtils {
    private final Random random = new Random();

    public int nextInt(int bound) {
        return random.nextInt(bound);
    }

    public double nextDouble(double bound) {
        return random.nextDouble() * bound;
    }

    public int nextInt(int lowerBound, int upperBound) {
        return nextInt(upperBound + 1 - lowerBound) + lowerBound;
    }

    public int nextWeightedInt(double[] weights) {
        double randInt = nextDouble(weights.length);
        int index;
        for (index = 0; randInt >= weights[index]; index++) {
            randInt -= weights[index];
        }
        return index;
    }
}

File: web/src/main/kotlin/uk/co/rafearnold/randomworkoutroutine/web/util/RandomUtils.kt
package uk.co.rafearnold.randomworkoutroutine.web.util

import kotlin.random.Random

/**
 * Returns a random integer between 0, inclusive, and the size of [weights], exclusive. The
 * probably of a given integer being selected is determined by the value at that index in
 * [weights]. Higher values are more likely to be selected than lower values.
 */
fun nextWeightedInt(weights: DoubleArray): Int {
    var randInt = Random.nextDouble(weights.sum())
    var index = 0
    while (randInt >= weights[index]) {
        randInt -= weights[index]
        index++
    }
    return index
}
